File: asys-runtime/javascript/channel.mjs
import { makeDirectory, fileMode } from './permissions.mjs';
// Ordered event streams between a component and the host over the shared root.
// Mirrors asys_runtime/channel.py: one file per event named by sequence number,
// published by an atomic link so numbers are claimed at publish time, and a
// per-direction cursor kept by the reader. No long-lived owner, nothing executed.
import { link, open, readdir, rm, stat, unlink } from 'node:fs/promises';
import { dirname, join, resolve } from 'node:path';
import { setTimeout } from 'node:timers/promises';
import { DatabaseSync } from 'node:sqlite';
import { readJSON, writeJSON } from './queue.mjs';

export const DIRECTIONS = ['in', 'out'];
export const POLL_MS = 50;
const WIDTH = 9;
const LIMIT = 10 ** WIDTH - 1;
const BYTES = 8 * 1024 * 1024;

export function channelRoot(root, name) {
  return join(resolve(root), 'channels', validName(name, 'Channel name'));
}

export function directionRoot(root, name, direction) {
  if (!DIRECTIONS.includes(direction)) throw new Error(`Channel direction must be one of ${DIRECTIONS.join(', ')}`);
  return join(channelRoot(root, name), direction);
}

class Stream {
  constructor(directory) { this.directory = resolve(directory); }
  async ready() {
    makeDirectory(this.directory, { recursive: true, mode: 0o700 });
    this.mode = (await stat(this.directory)).mode & 0o666;
    return this;
  }
  async exclusive(action) {
    await this.ready();
    // A short SQLite transaction supplies the same OS-backed lock to Python
    // and Node. The database stores no channel data. Never unlink it.
    const path = join(this.directory, '.write-lock.sqlite');
    await publishText(path, '', this.mode);
    const database = new DatabaseSync(path);
    try {
      const deadline = Date.now() + 30_000;
      for (;;) {
        try { database.exec('BEGIN EXCLUSIVE'); break; }
        catch (error) {
          if (![5, 6].includes(error.errcode) || Date.now() >= deadline) throw error;
          await setTimeout(10);
        }
      }
      const result = await action();
      database.exec('COMMIT');
      return result;
    } finally { database.close(); }
  }
  path(sequence) { return join(this.directory, `${String(sequence).padStart(WIDTH, '0')}.json`); }
  async sequences() {
    const sequences = [];
    for (const entry of await readdir(this.directory)) {
      const sequence = parseSequence(entry);
      if (sequence !== null) sequences.push(sequence);
    }
    return sequences.sort((a, b) => a - b);
  }
  async last() { const sequences = await this.sequences(); return sequences.length ? sequences.at(-1) : 0; }
  async event(sequence) { return validEvent(await readJSON(this.path(sequence)), this.directory, sequence); }
}

export class Writer extends Stream {
  constructor(directory) { super(directory); this.pending = Promise.resolve(); }
  async send(type, data = null) {
    validName(type, 'Event type');
    data = JSON.parse(JSON.stringify(data ?? null));
    const sending = this.pending.then(() => this.exclusive(async () => {
      const sequence = await this.last() + 1;
      if (sequence > LIMIT) throw new Error(`${this.directory}: channel sequence exhausted`);
      const event = { version: 1, sequence, type, time: timestamp(), data };
      if (!await publishJSON(this.path(sequence), event, this.mode)) throw new Error(`${this.directory}: concurrent publisher did not hold the channel lock`);
      return event;
    }));
    this.pending = sending.catch(() => {});
    return sending;
  }
}

export class Reader extends Stream {
  constructor(directory) { super(directory); this.cursorPath = join(this.directory, 'cursor.json'); }
  async cursor() {
    let value;
    try { value = await readJSON(this.cursorPath); }
    catch (error) { if (error.code === 'ENOENT') return 0; throw error; }
    if (!value || typeof value !== 'object' || !Number.isSafeInteger(value.after) || value.after < 0) throw new Error(`${this.cursorPath}: invalid cursor`);
    return value.after;
  }
  async advance(sequence) {
    if (!Number.isSafeInteger(sequence) || sequence < 0) throw new Error('Cursor must be a non-negative integer');
    await this.ready();
    await writeJSON(this.cursorPath, { version: 1, after: sequence }, { mode: this.mode });
  }
  async read(after, { limit } = {}) {
    await this.ready();
    if (after === undefined) after = await this.cursor();
    const events = [];
    for (const sequence of await this.sequences()) {
      if (sequence <= after) continue;
      if (limit !== undefined && events.length >= limit) break;
      try { events.push(await this.event(sequence)); }
      catch (error) { if (error.code !== 'ENOENT') throw error; }
    }
    return events;
  }
  // Yield events past `after` (default: the cursor) as they are published.
  // Ends after `timeoutMs` without a new event, or on `signal`. Never advances
  // the cursor: callers acknowledge with advance().
  async *follow(after, { signal, timeoutMs, pollMs = POLL_MS } = {}) {
    await this.ready();
    let position = after === undefined ? await this.cursor() : after;
    let deadline = timeoutMs === undefined ? Infinity : Date.now() + timeoutMs;
    for (;;) {
      signal?.throwIfAborted();
      const events = await this.read(position);
      if (events.length) {
        for (const event of events) { position = event.sequence; yield event; }
        deadline = timeoutMs === undefined ? Infinity : Date.now() + timeoutMs;
        continue;
      }
      if (Date.now() >= deadline) return;
      await setTimeout(pollMs, undefined, { signal });
    }
  }
  // Delete events up to and including `before` (default: the cursor). The
  // latest event always survives: it carries the high-water mark writers use.
  async prune(before) {
    if (before === undefined) before = await this.cursor();
    return this.exclusive(async () => {
      let removed = 0;
      for (const sequence of (await this.sequences()).slice(0, -1)) {
        if (sequence > before) break;
        await unlink(this.path(sequence)); removed++;
      }
      if (removed) await syncDirectory(this.directory);
      return removed;
    });
  }
}

function parseSequence(name) {
  const match = /^(\d{9})\.json$/u.exec(name);
  return match ? Number(match[1]) : null;
}

function validName(value, label) {
  if (typeof value !== 'string' || !/^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$/u.test(value)) throw new Error(`${label} is not a valid runtime name`);
  return value;
}

function validEvent(value, directory, sequence) {
  if (!value || typeof value !== 'object' || value.version !== 1 || !Number.isSafeInteger(value.sequence)
      || typeof value.type !== 'string' || typeof value.time !== 'string' || !('data' in value)) {
    throw new Error(`${directory}: invalid event ${sequence}`);
  }
  if (value.sequence !== sequence) throw new Error(`${directory}: event file ${sequence} claims sequence ${value.sequence}`);
  validName(value.type, 'Event type');
  return value;
}

function timestamp() {
  // Same shape as the Python side: ISO 8601 with microseconds and a UTC offset.
  return new Date().toISOString().replace(/Z$/u, '000+00:00');
}

// Create `path` with `value` atomically; false if it already exists.
async function publishJSON(path, value, mode) {
  const text = JSON.stringify(value, null, 2) + '\n';
  if (Buffer.byteLength(text) > BYTES) throw new Error(`${path}: JSON exceeds 8 MiB`);
  return publishText(path, text, mode);
}

async function publishText(path, text, mode) {
  const temporary = join(dirname(path), `.${path.split('/').at(-1)}.tmp.${process.pid}.${crypto.randomUUID()}`);
  try {
    const file = await open(temporary, 'wx', mode);
    try { await file.chmod(fileMode(path, mode)); await file.writeFile(text); await file.sync(); }
    finally { await file.close(); }
    try { await link(temporary, path); }
    catch (error) { if (error.code === 'EEXIST') return false; throw error; }
    await syncDirectory(dirname(path));
    return true;
  } finally { await rm(temporary, { force: true }); }
}

async function syncDirectory(path) {
  const directory = await open(path, 'r');
  try { await directory.sync(); } finally { await directory.close(); }
}

File: asys-runtime/asys_runtime/channel.py
"""Ordered event streams between a component and the host over the shared root.

A channel has two directions. Each direction is a directory of events, one file
per event, named by sequence number. Publishing is an atomic link, so sequence
numbers are claimed at publish time: a reader that sees event N knows every
event below N was published, even if later pruned. Readers keep a cursor in the direction
directory. A short transaction serializes publication and pruning; there is
no long-lived owner and nothing here executes anything.
"""

import json
import os
from contextlib import contextmanager
from pathlib import Path
import sqlite3
import time

from .files import read_json, sync_directory, timestamp, validate_name, write_json
from .permissions import mkdir

DIRECTIONS = ("in", "out")
POLL_SECONDS = 0.05
WIDTH = 9
LIMIT = 10 ** WIDTH - 1


def channel_root(root, name):
    validate_name("channel name", name)
    return Path(root).expanduser().resolve() / "channels" / name


def direction_root(root, name, direction):
    if direction not in DIRECTIONS:
        raise ValueError(f"channel direction must be one of {', '.join(DIRECTIONS)}")
    return channel_root(root, name) / direction


def event_path(directory, sequence):
    return Path(directory) / f"{sequence:0{WIDTH}d}.json"


def parse_sequence(filename):
    stem, dot, suffix = filename.partition(".")
    if dot and suffix == "json" and len(stem) == WIDTH and stem.isdigit():
        return int(stem)
    return None


def validate_event(value, directory, sequence=None):
    if (not isinstance(value, dict) or value.get("version") != 1 or type(value.get("sequence")) is not int
            or not isinstance(value.get("type"), str) or not isinstance(value.get("time"), str)
            or "data" not in value):
        raise ValueError(f"{directory}: invalid event {sequence}")
    if sequence is not None and value["sequence"] != sequence:
        raise ValueError(f"{directory}: event file {sequence} claims sequence {value['sequence']}")
    validate_name("event type", value["type"])
    return value


class Stream:
    """One direction of a channel."""

    def __init__(self, directory):
        self.directory = Path(directory).expanduser().resolve()
        mkdir(self.directory, parents=True, exist_ok=True)
        # Explicitly shared directions grant the same read/write permissions
        # to their files. The default direction remains private (0700/0600).
        self.mode = self.directory.stat().st_mode & 0o666

    @contextmanager
    def exclusive(self):
        # Python and Node both have SQLite in their standard libraries. Use
        # its OS-backed transaction lock only: all channel data stays in JSON.
        # Never unlink this file: a killed holder releases the lock itself.
        path = self.directory / ".write-lock.sqlite"
        publish_text(path, "", self.mode)
        connection = sqlite3.connect(path, timeout=30, isolation_level=None)
        try:
            connection.execute("BEGIN EXCLUSIVE")
            yield
            connection.execute("COMMIT")
        finally:
            connection.close()

    def sequences(self):
        return sorted(s for s in (parse_sequence(p.name) for p in self.directory.iterdir()) if s is not None)

    def last(self):
        sequences = self.sequences()
        return sequences[-1] if sequences else 0

    def event(self, sequence):
        return validate_event(read_json(event_path(self.directory, sequence)), self.directory, sequence)


class Writer(Stream):
    def send(self, event_type, data=None):
        validate_name("event type", event_type)
        json.dumps(data, allow_nan=False)
        with self.exclusive():
            sequence = self.last() + 1
            if sequence > LIMIT:
                raise ValueError(f"{self.directory}: channel sequence exhausted")
            event = {"version": 1, "sequence": sequence, "type": event_type, "time": timestamp(), "data": data}
            if not publish_json(event_path(self.directory, sequence), event):
                raise ValueError(f"{self.directory}: concurrent publisher did not hold the channel lock")
            return event


class Reader(Stream):
    def __init__(self, directory):
        super().__init__(directory)
        self.cursor_path = self.directory / "cursor.json"

    @property
    def cursor(self):
        try:
            value = read_json(self.cursor_path)
        except FileNotFoundError:
            return 0
        if not isinstance(value, dict) or type(value.get("after")) is not int or value["after"] < 0:
            raise ValueError(f"{self.cursor_path}: invalid cursor")
        return value["after"]

    def advance(self, sequence):
        if type(sequence) is not int or sequence < 0:
            raise ValueError("cursor must be a non-negative integer")
        write_json(self.cursor_path, {"version": 1, "after": sequence}, mode=self.mode)

    def read(self, after=None, *, limit=None):
        after = self.cursor if after is None else after
        events = []
        for sequence in self.sequences():
            if sequence <= after:
                continue
            if limit is not None and len(events) >= limit:
                break
            try:
                events.append(self.event(sequence))
            except FileNotFoundError:
                continue  # pruned between listing and reading
        return events

    def follow(self, after=None, *, timeout=None, poll=POLL_SECONDS):
        """Yield events past `after` (default: the cursor) as they are published.

        Stops after `timeout` seconds without a new event; runs forever if None.
        The cursor is not advanced: callers acknowledge with advance()."""
        position = self.cursor if after is None else after
        deadline = None if timeout is None else time.monotonic() + timeout
        while True:
            events = self.read(position)
            if events:
                for event in events:
                    position = event["sequence"]
                    yield event
                deadline = None if timeout is None else time.monotonic() + timeout
                continue
            if deadline is not None and time.monotonic() >= deadline:
                return
            time.sleep(poll)

    def prune(self, before=None):
        """Delete events up to and including `before` (default: the cursor).

        The latest event always survives: it carries the sequence high-water
        mark that writers derive their next number from."""
        before = self.cursor if before is None else before
        with self.exclusive():
            removed = 0
            sequences = self.sequences()
            for sequence in sequences[:-1]:
                if sequence > before:
                    break
                event_path(self.directory, sequence).unlink()
                removed += 1
            if removed:
                sync_directory(self.directory)
            return removed


def publish_json(path, value):
    """Create `path` with `value` atomically; return False if it already exists."""
    text = json.dumps(value, ensure_ascii=False, allow_nan=False, indent=2) + "\n"
    if len(text.encode("utf-8")) > 8 * 1024 * 1024:
        raise ValueError(f"{path}: JSON exceeds 8 MiB")
    return publish_text(path, text, Path(path).parent.stat().st_mode & 0o666)


def publish_text(path, text, mode):
    path = Path(path)
    tmp = path.with_name(f".{path.name}.tmp.{os.getpid()}.{os.urandom(6).hex()}")
    try:
        with tmp.open("x", encoding="utf-8") as stream:
            os.chmod(tmp, mode)
            stream.write(text)
            stream.flush()
            os.fsync(stream.fileno())
        try:
            os.link(tmp, path)
        except FileExistsError:
            return False
        sync_directory(path.parent)
        return True
    finally:
        tmp.unlink(missing_ok=True)
